### PR TITLE
fix bug that undefined internal is a warning only for -pedantic-errors

### PR DESCRIPTION
**Pull request by ccrownhill**
This fixes issue #39558 which mentions the problem when compiling the following code with `-pedantic-errors` flag (it also mentions `-Wall -Wextra` but those shouldn't change the output, which is also the case in GCC):

```c
static void f();

int main()
{
    f;
}
```

Clang only outputs an `undefined-internal` warning on the first line, but according to 6.9/3

```
"... Moreover, if an identifier declared with internal linkage is used in an
 expression (other than as a part of the operand of a sizeof or _Alignof
 operator whose result is an integer constant), there shall be exactly one
 external definition for the identifier in the translation unit."
```

this should be illegal and hence an error.

I fixed this by changing the warning type from `Warning` to `ExtWarn` and by adding a suitable test.

### CODE REVIEW

**Review by Sirraide — APPROVED** [review]
Thanks for the fix! 

The changes LGTM, but you should also add a release note about this (in `clang/docs/ReleaseNotes.rst`, probably in the ‘Improvements to Clang’s Diagnostics’ section).

**Comment by ccrownhill** [comment]
Ok I just added a commit with the addition to `clang/docs/ReleaseNotes.rst`.

**Review by AaronBallman — COMMENTED** [review]
Thank you for working on this! C23 changed the normative wording here somewhat, so I think we should add additional test coverage. Specifically:

Moreover, if an identifier declared with internal linkage is used in an expression
there shall be exactly one external definition for the identifier in the translation unit, unless it is:
— part of the operand of a sizeof operator whose result is an integer constant;
— part of the operand of an alignof operator whose result is an integer constant;
— part of the controlling expression of a generic selection;
— part of the expression in a generic association that is not the result expression of its generic
selection;
— or, part of the operand of any typeof operator whose result is not a variably modified type.

How about adding those bullets as test cases to make sure we don't diagnose them?

**Comment by ccrownhill** [comment]
I added three more tests covering these points:
— part of the operand of a sizeof operator whose result is an integer constant;
— part of the controlling expression of a generic selection;
— or, part of the operand of any typeof operator whose result is not a variably modified type.

I did not see how to add tests for the following since they require just a type which I thought to be possible to achieve
with an undefined internal since for a type the first declaration is the definition, e.g. with typedef.
— part of the operand of an alignof operator whose result is an integer constant;
— part of the expression in a generic association that is not the result expression of its generic
selection; (*this one is also a type since the association map contains only pairs of type names and result expressions*)

Is there a way to test undefined internals for these cases?

**Comment by AaronBallman** [comment]
> I did not see how to add tests for the following since they require just a type which I thought to be possible to achieve with an undefined internal since for a type the first declaration is the definition, e.g. with typedef.
> — part of the operand of an alignof operator whose result is an integer constant; 
> — part of the expression in a generic association that is not the result expression of its generic selection; (_this one is also a type since the association map contains only pairs of type names and result expressions_)
> 
> Is there a way to test undefined internals for these cases?

What I had in mind for those was:
```
static void f();
int i = _Alignof(f);
int j = _Generic(&f, void (*)(void) : 1, default : 0);
```
https://godbolt.org/z/6dzYn4jn9

**Comment by ccrownhill** [comment]
Ah I see. Thank you for the suggestions!
I also tried the same for `_Alignof` but was put off by the errors generated by the `-pedantic-errors` flag.
But I guess I can just check that I don't get the `undefined-internal` error.

For the other two I had something similar but that still misses this point:
— part of the expression in a generic association that is not the result expression of its generic
selection
But that might also be because I don't understand this point well enough. I don't see which expression is there that is not the controlling expression and not a result expression.

Also, I was thinking that it would be best to group all tests into one file since they all test just one feature.
What do you think of that?

**Comment by ccrownhill** [comment]
Since one error message will occlude others I had to split the files as I did in the latest commit.
I have added one for the `_Alignof` case checking for the errors I was getting in godbolt while also checking that no `undefined-internal` error occurred.
Lastly, I added the one illegal use case of `_Generic` from @AaronBallman 's suggestions.
I think this should cover everything.
What do you think?

**Comment by AaronBallman** [comment]
> For the other two I had something similar but that still misses this point:
> — part of the expression in a generic association that is not the result expression of its generic
> selection
> But that might also be because I don't understand this point well enough. I don't see which expression is there that is not the controlling expression and not a result expression.

Given: `_Generic(foo, bar : baz)`, `foo` is the controlling expression, `bar` is the generic association, and assuming `foo` is of type `bar`, then `baz` is the result expression of the generic selection.

**Comment by ccrownhill** [comment]
Oh that makes sense so

```
static void f();
void *k = _Generic(&f, void (*)(void) : 0, default : f);
```

should be ok since it is not in the result expression?

File: clang/test/Sema/undefined-internal-alignof.c [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s -pedantic-errors
```
Comment (AaronBallman) [inline clang/test/Sema/undefined-internal-alignof.c:1]:
```suggestion
// RUN: %clang_cc1 -fsyntax-only -verify %s -Wno-pointer-arith -pedantic-errors
```

File: clang/test/Sema/undefined-internal-generic-selection.c [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s -pedantic-errors
+
+static void *f(void); // expected-error {{function 'f' has internal linkage but is not defined}}
+
+int main(void)
+{
+    int j = _Generic(&f, void *(*)(void): 10, default: 20);
+    // expected-no-diagnostic@-1
+
+    void *k = _Generic(&f, void *(*)(void): f(), default: 20);
```
Comment (AaronBallman) [inline clang/test/Sema/undefined-internal-generic-selection.c:10]:
```suggestion
    int j = _Generic(&f, void (*)(void): 0);
    void *k = _Generic(&f, void (*)(void): f, default: 0); // expected-note {{used here}}
```

File: clang/test/Sema/undefined-internal-sizeof.c [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s -pedantic-errors
+
+// expected-no-diagnostics
+
+static int f(void);
+
+int main(void)
+{
+    int x = sizeof f();
```
Comment (AaronBallman) [inline clang/test/Sema/undefined-internal-sizeof.c:9]:
```suggestion
    int x = sizeof(f);
```

File: clang/test/Sema/undefined-internal-sizeof.c [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s -pedantic-errors
```
Comment (AaronBallman) [inline clang/test/Sema/undefined-internal-sizeof.c:1]:
```suggestion
// RUN: %clang_cc1 -fsyntax-only -verify %s -Wno-pointer-arith -pedantic-errors
```

File: clang/test/Sema/undefined-internal-typeof-c23.c [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s -std=c23 -pedantic-errors
+
+// expected-no-diagnostics
+
+static int f(void);
+
+int main(void)
+{
+    typeof(f()) x;
```
Comment (AaronBallman) [inline clang/test/Sema/undefined-internal-typeof-c23.c:9]:
```suggestion
    typeof(&f) x;
```

**Comment by ccrownhill** [comment]
Sorry, my mistake. This doesn't work for `default` but the problem is that in this example

```
static void *f(void);
void *k = _Generic(&f, void *(*)(void) : 0, int (*)(void) : f());
```

the `undefined-internal` error is still raised even though it is not in the result expression anymore.

**Review by AaronBallman — COMMENTED** [review]
I think the test cases should all be combined into a single file -- we don't stop issuing diagnostics after the first one, so it should be fine to put them all together into one file.

**Comment by AaronBallman** [comment]
> Oh that makes sense so
> 
> ```
> static void f();
> void *k = _Generic(&f, void (*)(void) : 0, default : f);
> ```
> 
> should be ok since it is not in the result expression?

Correct, that's a good test case to add! Similarly, it would be good to add one like:
```
void *k = _Generic(&f, int (*)(void) : 0, default : f);
```
to show that it is diagnosed because it's used in the result expression.

**Comment by ccrownhill** [comment]
Thank you very much for all the feedback:) I'll combine them into a file now.

**Comment by ccrownhill** [comment]
> Correct, that's a good test case to add!

The only problem is that currently, the behavior of Clang is incorrect for the case

```
static void *f(void);
void *k = _Generic(&f, void *(*)(void) : 0, default : f());
```
because it gives an error even though it should not.
But I can have a look into how to fix that.

**Comment by AaronBallman** [comment]
> > Correct, that's a good test case to add!
> 
> The only problem is that currently, the behavior of Clang is incorrect for the case
> 
> ```
> static void *f(void);
> void *k = _Generic(&f, void *(*)(void) : 0, default : f());
> ```
> 
> because it gives an error even though it should not. But I can have a look into how to fix that.

If you look into it and it seems challenging to resolve, I think we can leave a FIXME comment next to the test case saying "this code is incorrectly rejected and should be accepted", file an issue in GitHub about it, and then land these changes because they're definitely an incremental improvement over what we have today.

**Comment by ccrownhill** [comment]
> If you look into it and it seems challenging to resolve, I think we can leave a FIXME comment next to the test case saying "this code is incorrectly rejected and should be accepted", file an issue in GitHub about it, and then land these changes because they're definitely an incremental improvement over what we have today.

Ok, I'll have a look after adding the above test cases.
Should I include the test that will break it for that already?

And also since the `typeof` test only works for C23 should I add it to the single test file and compile with `-std=c23`
or leave it in a separate test file?

**Comment by AaronBallman** [comment]
> > If you look into it and it seems challenging to resolve, I think we can leave a FIXME comment next to the test case saying "this code is incorrectly rejected and should be accepted", file an issue in GitHub about it, and then land these changes because they're definitely an incremental improvement over what we have today.
> 
> Ok, I'll have a look after adding the above test cases. Should I include the test that will break it for that already?

Yeah, go ahead and add the test, and add whatever `// expected-` comment to get the test to pass along with the FIXME comment.

> And also since the `typeof` test only works for C23 should I add it to the single test file and compile with `-std=c23` or leave it in a separate test file?

I'd probably spell it `__typeof__` and leave the language standard unspecified, that's reasonable because it's a synonym for `typeof`.

**Comment by ccrownhill** [comment]
Ok, that's done with the FIXME comment for now.
Maybe it is only because this is my first time contributing and I am not too familiar with the codebase but I couldn't see a trivial fix for the issue.
For now I think it's best to land these changes first.
I would just add an issue about it.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by AaronBallman** [comment]
I filed https://github.com/llvm/llvm-project/issues/98504 to track the remaining issue.

**Comment by AaronBallman** [comment]
I've reverted the changes in 9f283bfb9e66d57fefeb3a28318ec40081fa9a24 because of the bot breakage. Oddly, I am seeing a different failure on Windows locally:
```
FAIL: Clang :: Sema/undefined-internal-basic.c (1 of 1)
******************** TEST 'Clang :: Sema/undefined-internal-basic.c' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem F:\source\llvm-project\llvm\out\build\x64-Debug\lib\clang\19\include -nostdsysteminc -fsyntax-only -verify F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c -Wno-pointer-arith -Wno-gnu-alignof-expression -Wno-unused -pedantic-errors
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'F:\source\llvm-project\llvm\out\build\x64-Debug\lib\clang\19\include' -nostdsysteminc -fsyntax-only -verify 'F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c' -Wno-pointer-arith -Wno-gnu-alignof-expression -Wno-unused -pedantic-errors
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 3: function 'a' has internal linkage but is not defined
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 4: function 'b' has internal linkage but is not defined
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 5: function 'c' has internal linkage but is not defined
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 6: function 'd' has internal linkage but is not defined
# | error: 'expected-warning' diagnostics seen but not expected:
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 3: function 'a' has internal linkage but is not defined
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 4: function 'b' has internal linkage but is not defined
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 5: function 'c' has internal linkage but is not defined
# |   File F:\source\llvm-project\clang\test\Sema\undefined-internal-basic.c Line 6: function 'd' has internal linkage but is not defined
# | 8 errors generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..
********************
Failed Tests (1):
  Clang :: Sema/undefined-internal-basic.c
```
so something has gone amiss...

**Comment by AaronBallman** [comment]
Okay, the Windows issue seems to have gone away with a full rebuild, which makes *way* more sense to me. :-) So I think it's just the `llvm-clang-x86_64-sie-ubuntu-fast` issue that needs to be resolved.

**Comment by AaronBallman** [comment]
I think adding `-std=c11` to the `RUN` line should help; it at least removes the "is a C11 extension" diagnostics. (I guess this builder isn't defaulting to C17 for some reason?)

**Comment by AaronBallman** [comment]
I've relanded the changes with a fix here in commit 0171e23647d58f65c1cddbc22e16fda2bafa6e85, hopefully that sticks!

**Comment by AaronBallman** [comment]
Yup, that resolved the issue: https://lab.llvm.org/buildbot/#/builders/144/builds/2082 -- not certain why that builder has a different default value for `-std` though. Clang defaults to C17, but that builder seems to be defaulting to C99.

**Comment by dyung** [comment]
> Yup, that resolved the issue: https://lab.llvm.org/buildbot/#/builders/144/builds/2082 -- not certain why that builder has a different default value for `-std` though. Clang defaults to C17, but that builder seems to be defaulting to C99.

This bot targets the PS4 which has kept an older version of the standard as the default for compatibility reasons.

**Comment by AaronBallman** [comment]
> > Yup, that resolved the issue: https://lab.llvm.org/buildbot/#/builders/144/builds/2082 -- not certain why that builder has a different default value for `-std` though. Clang defaults to C17, but that builder seems to be defaulting to C99.
> 
> This bot targets the PS4 which has kept an older version of the standard as the default for compatibility reasons.

Ahhhhh! I remembered that the PS used older versions of the standard, but I didn't realize that was *this* bot. "SIE" -- Sony Interactive Entertainment... now I get it. Thank you! :-)

**Comment by AaronBallman** [comment]
> Thank you very much for fixing that:)

Happy to help, thank you for your contribution!

**Pull request merged**